spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteMergeIntoTable.scala

```diff
@@ -0,0 +1,267 @@
+    // use full outer join in all other cases, unmatched source rows may be needed
+    // disable broadcasts for the target table to perform the cardinality check
+    val joinType = if (notMatchedActions.isEmpty) LeftOuter else FullOuter
+    val joinHint = JoinHint(leftHint = Some(HintInfo(Some(NO_BROADCAST_HASH))), rightHint = None)
+    val joinPlan = Join(targetTableProj, sourceTableProj, joinType, Some(cond), joinHint)
+
+    val rowIdAttr = ExtendedV2ExpressionUtils.resolveRef[AttributeReference](
+      FieldReference(ROW_ID),
+      joinPlan)
+    val rowFromSourceAttr = ExtendedV2ExpressionUtils.resolveRef[AttributeReference](
+      FieldReference(ROW_FROM_SOURCE),
+      joinPlan)
+    val rowFromTargetAttr = ExtendedV2ExpressionUtils.resolveRef[AttributeReference](
+      FieldReference(ROW_FROM_TARGET),
+      joinPlan)
+
+    // add an extra matched action to output the original row if none of the actual actions matched
+    // this is needed to keep target rows that should be copied over as we are working with groups
+    val mergeRowsParams = MergeRowsParams(
+      isSourceRowPresent = IsNotNull(rowFromSourceAttr),
+      isTargetRowPresent = IsNotNull(rowFromTargetAttr),
+      matchedConditions = matchedActions.map(actionCondition) :+ TrueLiteral,
+      matchedOutputs = matchedActions.map(output(_, metadataAttrs)) :+ Some(readAttrs),
+      notMatchedConditions = notMatchedActions.map(actionCondition),
+      notMatchedOutputs = notMatchedActions.map(output(_, metadataAttrs)),
+      targetOutput = readAttrs,
+      joinedAttributes = joinPlan.output,
+      rowIdAttrs = Seq(rowIdAttr),
+      performCardinalityCheck = isCardinalityCheckNeeded(matchedActions),
+      emitNotMatchedTargetRows = true)
+    val mergeRows = buildMergeRows(mergeRowsParams, readAttrs, joinPlan)
+
+    // build a plan to replace read groups in the table
+    val writeRelation = relation.copy(table = table)
+    ReplaceData(writeRelation, mergeRows, relation)
+  }
+
+  private def actionCondition(action: MergeAction): Expression = {
+    action.condition.getOrElse(TrueLiteral)
+  }
+
+  private def output(
+      clause: MergeAction,
+      metadataAttrs: Seq[Attribute]): Option[Seq[Expression]] = {
+
+    clause match {
+      case u: UpdateAction =>
+        Some(u.assignments.map(_.value) ++ metadataAttrs)
+
+      case _: DeleteAction =>
+        None
+
+      case i: InsertAction =>
+        Some(i.assignments.map(_.value) ++ metadataAttrs.map(attr => Literal(null, attr.dataType)))
+
+      case other =>
+        throw new AnalysisException(s"Unexpected action: $other")
+    }
+  }
+
+  private def buildMergeRows(
+      params: MergeRowsParams,
+      attrs: Seq[Attribute],
+      joinPlan: LogicalPlan): MergeRows = {
+
+    val outputs = params.matchedOutputs.flatten ++ params.notMatchedOutputs.flatten
+    assert(outputs.nonEmpty, "must be at least one output")
```

Comment:
Should this be an analysis error instead? Or is there no valid SQL that could produce this case?

Reply:
Yeah, I think it is not possible to hit this case as the parser would complain. Just added to be safe.

Reply:
No longer applies.